libs/utils/utils.js

```diff
@@ -721,13 +721,47 @@ function decorateSections(el, isDoc) {
   return [...el.querySelectorAll(selector)].map(decorateSection);
 }
 
-function decorateFooterPromo(config) {
-  const footerPromo = getMetadata('footer-promo-tag');
-  if (!footerPromo) return;
-  const href = `${config.locale.contentRoot}/fragments/footer-promos/${footerPromo}`;
-  const para = createTag('p', {}, createTag('a', { href }, href));
-  const section = createTag('div', null, para);
-  document.querySelector('main > div:last-of-type').insertAdjacentElement('afterend', section);
+async function getFooterPromoByTag(contentRoot) {
+  const NAME = 'Name';
+  const FOOTER_PROMO_LINK = 'Footer Promo Link';
+  const taxonomyUrl = `${contentRoot}/taxonomy.json`;
+  const tagsOnPage = [...document.head.querySelectorAll('meta[property="article:tag"]')].map((el) => el.content);
+
+  if (!tagsOnPage.length) return false;
+
+  try {
+    const resp = await fetch(taxonomyUrl);
+    if (!resp.ok) return false;
+    const { data } = await resp.json();
+    const primaryTag = data.find((tag) => {
+      const name = tag[NAME].split('|').pop().trim();
+      return tagsOnPage.includes(name) && tag[FOOTER_PROMO_LINK];
+    });
+    if (primaryTag) return primaryTag[FOOTER_PROMO_LINK];
+  } catch (error) {
+    window.lana.log(`Footer Promo - Taxonomy error: ${error}`);
+    return false;
+  }
+  return false;
+}
+
+export async function decorateFooterPromo({ locale: { contentRoot } }) {
+  const footerPromoMeta = getMetadata('footer-promo-tag');
+  const tagBased = getMetadata('tag-based-footer-promo');
+  if ((!footerPromoMeta || footerPromoMeta === 'off') && tagBased !== 'on') return;
+  let href;
```

Comment:
As this is becoming increasingly larger, let's turn this into a separate function and dynamically load it.
```suggestion
export async function decorateFooterPromo({ locale: { contentRoot } }) {
  const footerPromoMeta = getMetadata('footer-promo-tag');
  const tagBased = getMetadata('tag-based-footer-promo');
  if ((!footerPromoMeta || footerPromoMeta === 'off') && tagBased !== 'on') return;
  const { decorateFooterPromo } = await import('../features/footerPromo.js');
  decorateFooterPromo({ locale: { contentRoot } });
```